airbyte-integrations/bases/base-python/base_python/sdk/streams/core.py

```diff
@@ -0,0 +1,96 @@
+import inspect
+from abc import ABC, abstractmethod
+from typing import Mapping, Any, Iterable, List, Union
+from base_python.logger import AirbyteLogger
+from base_python.schema_helpers import ResourceSchemaLoader
+
+
+def package_name_from_class(cls: object) -> str:
+    """Find the package name given a class name"""
+    module = inspect.getmodule(cls)
+    return module.__name__.split(".")[0]
+
+
+class Stream(ABC):
+    # Use self.logger in subclasses to log any messages
+    logger = AirbyteLogger()  # TODO use native "logging" loggers with custom handlers
+
+    @property
+    def name(self) -> str:
+        """
+        :return: Stream name. By default this is the implementing class name, but it can be overridden as needed.
+        """
+        return self.__class__.__name__
+
+    @abstractmethod
+    def read_stream(self, stream_state: Mapping[str, Any] = {}) -> Iterable[Mapping[str, Any]]:
+        """
+        This method should be overridden by subclasses
+
+        :param stream_state: State dict to use when extracting records.
+        :return: A generator which yields all the records in this stream. Each record is a dict from properties to their values matching the schema
+         of the stream.
+        """
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        :return: A dict of the JSON schema representing this stream.
+
+        The default implementation of this method looks for a JSONSchema file with the same name as this stream's "name" property.
+        Override as needed.
+        """
+        # TODO show an example of using pydantic to define the JSON schema, or reading an OpenAPI spec
+        # TODO change to snakecase by default
+        print(self.__class__)
+        print(self.name)
+        return ResourceSchemaLoader(package_name_from_class(self.__class__)).get_schema(self.name)
+
+
+class IncrementalStream(Stream, ABC):
```

Comment:
The name is actually a bit misleading. It is not really an IncrementalStream but a stream that supports Incremental

Reply:
I'm not sure I understand your comment @michel-tricot. could you explain it a bit more?

Reply:
Yes, I don't believe we can actually split the abstraction this way. Just look at read that takes in a state wether it is incremental or not. 

I think we should just put all of that into stream and have sensible defaults. 

It will cleanup all the isinstance code or distinction between incremental and not.

Reply:
merged IncrementalStream with Stream